drivers/serial/uart_hummingbird.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2021, TOKITA Hiroshi
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuclei_hummingbird_uart
+
+#include <drivers/uart.h>
+
+#define _AC(X, Y)        (X##Y)
+
+#define GPIO_CTRL_ADDR          _AC(0x10012000, UL)
+#define IOF0_UART0_MASK         _AC(0x00030000, UL)
+#define _REG32(p, i)            (*(volatile uint32_t *) ((p) + (i)))
+#define GPIO_REG(offset)        _REG32(GPIO_CTRL_ADDR, offset)
+
+#define GPIO_IOF_EN     (0x38)
+#define GPIO_IOF_SEL    (0x3C)
+
+/* TXCTRL register */
+#define UART_TXEN               0x1
+#define UART_TXWM(x)            (((x) & 0xffff) << 16)
+
+/* RXCTRL register */
+#define UART_RXEN               0x1
+#define UART_RXWM(x)            (((x) & 0xffff) << 16)
+
+/* IP register */
+#define UART_IP_TXWM            0x1
+#define UART_IP_RXWM            0x2
+
+#define UART_TXFIFO_FULL        (1 << 31)
+#define UART_RXFIFO_EMPTY       (1 << 31)
+
+#define UART_TXCTRL_TXCNT_OFS   (16)
+#define UART_TXCTRL_TXCNT_MASK  (0x7 << UART_TXCTRL_TXCNT_OFS)
+#define UART_TXCTRL_TXEN_OFS    (0)
+#define UART_TXCTRL_TXEN_MASK   (0x1 << UART_TXCTRL_TXEN_OFS)
+#define UART_TXCTRL_NSTOP_OFS   (1)
+#define UART_TXCTRL_NSTOP_MASK  (0x1 << UART_TXCTRL_TXEN_OFS)
+
+#define UART_RXCTRL_RXCNT_OFS   (16)
+#define UART_RXCTRL_RXCNT_MASK  (0x7 << UART_RXCTRL_RXCNT_OFS)
+#define UART_RXCTRL_RXEN_OFS    (0)
+#define UART_RXCTRL_RXEN_MASK   (0x1 << UART_RXCTRL_RXEN_OFS)
+
+#define UART_IE_TXIE_OFS        (0)
+#define UART_IE_TXIE_MASK       (0x1 << UART_IE_TXIE_OFS)
+#define UART_IE_RXIE_OFS        (1)
+#define UART_IE_RXIE_MASK       (0x1 << UART_IE_RXIE_OFS)
+
+#define UART_IP_TXIP_OFS        (0)
+#define UART_IP_TXIP_MASK       (0x1 << UART_IP_TXIP_OFS)
+#define UART_IP_RXIP_OFS        (1)
+#define UART_IP_RXIP_MASK       (0x1 << UART_IP_RXIP_OFS)
+
+#define CFG2UART(cfg) ((UART_TypeDef *)((const struct hummingbird_uart_config *)dev->config)->reg)
+
+
+/**
+ * @brief UART
+ */
+typedef struct {
+	volatile uint32_t TXFIFO;
+	volatile uint32_t RXFIFO;
+	volatile uint32_t TXCTRL;
+	volatile uint32_t RXCTRL;
+	volatile uint32_t IE;
+	volatile uint32_t IP;
+	volatile uint32_t DIV;
+} UART_TypeDef;
```

Comment:
If I remember correctly, declaring new typedefs is against the Zephyr coding standard.

Reply:
Removed typedef and 
Replace UART_TypeDef with struct UART_TypeDef whole file.